pkgs/top-level/all-packages.nix

```diff
@@ -16245,15 +16245,57 @@ with pkgs;
   gcc-arm-embedded-12 = callPackage ../development/compilers/gcc-arm-embedded/12 { };
   gcc-arm-embedded = gcc-arm-embedded-12;
 
-  # It would be better to match the default gcc so that there are no linking errors
-  # when using C/C++ libraries in D packages, but right now versions >= 12 are broken.
-  gdc = gdc11;
+  # Has to match the default gcc so that there are no linking errors when
+  # using C/C++ libraries in D packages
+  gdc = gdc13;
   gdc11 = wrapCC (gcc11.cc.override {
     name = "gdc";
-    langCC = false;
-    langC = false;
+    langCC = true; # required to bootstrap newer gdc
+    langC = true;
     langD = true;
     profiledCompiler = false;
+    gdc = null; # gdc>=12 requires gdc to bootstrap
+  } // {
+    # GDC doesn't support aarch64-darwin yet
+    meta.broken = stdenv.hostPlatform.isDarwin && stdenv.hostPlatform.isAarch64;
+  });
+
+  gdc12 = wrapCC (gcc12.cc.override {
+    name = "gdc";
+    langCC = true;
+    langC = true;
+    langD = true;
+    profiledCompiler = false;
+    gdc = gdc11; # gdc>=12 requires gdc to bootstrap
+    stdenv =
+      if stdenv.hostPlatform == stdenv.targetPlatform
+         && stdenv.buildPlatform == stdenv.hostPlatform
+         && stdenv.buildPlatform.isDarwin
+         && stdenv.buildPlatform.isx86_64
+      then overrideCC stdenv gdc11
+      else stdenv;
+  } // {
+    # GDC doesn't support aarch64-darwin yet
+    meta.broken = stdenv.hostPlatform.isDarwin && stdenv.hostPlatform.isAarch64;
+  });
+
+  gdc13 = wrapCC (gcc13.cc.override {
+    name = "gdc";
+    langCC = true;
+    langC = true;
+    langD = true;
+    profiledCompiler = false;
+    gdc = gdc11; # gdc>=12 requires gdc to bootstrap
+    stdenv =
+      if stdenv.hostPlatform == stdenv.targetPlatform
+         && stdenv.buildPlatform == stdenv.hostPlatform
+         && stdenv.buildPlatform.isDarwin
+         && stdenv.buildPlatform.isx86_64
+      then overrideCC stdenv gdc11
+      else stdenv;
+  } // {
+    # GDC doesn't support aarch64-darwin yet
+    meta.broken = stdenv.hostPlatform.isDarwin && stdenv.hostPlatform.isAarch64;
```

Comment:
This doesn't belong in `all-packges.nix`.  Please move it into `pkgs/development/compilers/gcc`, and please express it using conditionals instead of copy-paste-tweak.